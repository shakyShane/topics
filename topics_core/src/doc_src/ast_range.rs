pub trait AstRangeImpl {
    fn ast_range(&self) -> AstRange;
}

///
/// An AST (tree-like) range selector
///
/// ```rust
/// # fn main() -> anyhow::Result<()> {
/// # use comrak::Arena;
/// # use topics_core::doc_src::MdElements;
/// let input = "# heading";
/// let arena = Arena::new();
/// let md_elements = MdElements::new(input, &arena);
///
/// // tuple of (vec![0], 1) is path + len
/// let html = md_elements.as_html((vec![0], 1));
/// assert_eq!(html, String::from("<h1>heading</h1>\n"));
/// # Ok(())
/// # }
/// ```
#[derive(Debug, Clone, Default)]
pub struct AstRange {
    pub ast_path: Vec<usize>,
    pub ast_len: usize,
}

impl AstRange {
    pub fn range(path: &[usize], len: usize) -> Self {
        Self {
            ast_path: path.to_vec(),
            ast_len: len,
        }
    }
}

impl AstRangeImpl for AstRange {
    fn ast_range(&self) -> AstRange {
        self.clone()
    }
}

impl AstRangeImpl for (Vec<usize>, usize) {
    fn ast_range(&self) -> AstRange {
        AstRange {
            ast_path: self.0.to_vec(),
            ast_len: self.1,
        }
    }
}

impl AstRangeImpl for (&'_ Vec<usize>, usize) {
    fn ast_range(&self) -> AstRange {
        AstRange {
            ast_path: self.0.to_vec(),
            ast_len: self.1,
        }
    }
}
