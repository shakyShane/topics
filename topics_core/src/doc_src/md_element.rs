use crate::doc::DocResult;

pub struct MdElements {
    elements: Vec<Element>,
}

#[derive(Clone, Debug)]
pub enum Element {
    Heading {
        level: usize,
        content: String,
    },
    List {
        items: Vec<ListItem>,
    },
    Paragraph {
        content: String,
    },
    Text {
        content: String,
    },
    CodeBlock {
        params: Option<String>,
        content: String,
    },
}

#[derive(Clone, Debug)]
pub struct ListItem(Vec<Element>);
