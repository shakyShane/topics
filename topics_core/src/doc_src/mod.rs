pub mod doc_src;
pub mod yaml_doc_src;

pub use doc_src::*;
pub use yaml_doc_src::*;
