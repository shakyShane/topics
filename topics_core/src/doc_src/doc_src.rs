use crate::context::Context;
use crate::doc::DocResult;
use crate::doc_src::{MdDocSource, TomlDocSource, YamlDocSource};

use multi_doc::SingleDoc;
use std::path::PathBuf;

pub trait DocSrcImpl: Sized {
    fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self>;
}

#[derive(Debug, Clone)]
pub enum DocSource {
    Yaml(YamlDocSource),
    Toml(TomlDocSource),
    Md(MdDocSource),
}

impl DocSource {
    pub fn file(&self) -> Option<PathBuf> {
        match self {
            DocSource::Yaml(yaml_doc) => yaml_doc.input_file.clone(),
            DocSource::Toml(toml_doc) => toml_doc.input_file.clone(),
            DocSource::Md(md_doc) => md_doc.input_file.clone(),
        }
    }
    pub fn content(&self) -> &str {
        match self {
            DocSource::Yaml(yaml_doc) => yaml_doc.file_content.as_str(),
            DocSource::Toml(toml_doc) => toml_doc.file_content.as_str(),
            DocSource::Md(md_doc) => md_doc.file_content.as_str(),
        }
    }
    pub fn yaml(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Yaml(YamlDocSource::from_path_buf(&pb, ctx)?))
    }
    pub fn toml(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Toml(TomlDocSource::from_path_buf(&pb, ctx)?))
    }
    pub fn md(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Md(MdDocSource::from_path_buf(&pb, ctx)?))
    }
}

impl Default for DocSource {
    fn default() -> Self {
        Self::Yaml(YamlDocSource::default())
    }
}
