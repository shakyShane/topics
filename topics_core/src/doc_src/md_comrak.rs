use std::fmt;
use std::fmt::{Debug, Formatter};

use comrak::nodes::{Ast, AstNode, NodeCodeBlock, NodeHeading, NodeValue};

use crate::doc_src::ast_range::AstRange;
use crate::doc_src::{parse_inline_kind, MdElements};
use crate::items::{Command, Instruction, Item, ItemWrap, LineMarker};
use comrak::arena_tree::Node;
use std::cell::RefCell;

pub(crate) fn process_node<'a>(node: &'a AstNode<'a>, path: &mut Vec<usize>) -> Vec<Item> {
    let mut kind: Option<Item> = None;
    let mut items: Vec<Item> = vec![];
    let first = node.children().take(1).nth(0);
    // let mut other = node.children().skip(1);

    // process the very first item, expected to be a heading, but maybe not?
    if let Some(node) = first {
        let ast = node.data.borrow();
        match &ast.value {
            NodeValue::Heading(NodeHeading { level: 1, .. }) => {
                let start_line = ast.start_line;
                let t = collect_single_line_text(node);
                let item = parse_inline_kind(&t);
                kind = item;
                if let Some(item) = kind.as_mut() {
                    item.set_line_start(start_line)
                }
            }
            _ => {
                todo!("handle other 'first' elements?")
            }
        }
    }

    if let Some(Item::Instruction(inst)) = kind.as_mut() {
        let Instruction { name: _, ast_range } = inst;
        *ast_range = AstRange::range(&path, node.children().count());
    }

    // todo: probably select many command, for MVP just select the first one seen
    if let Some(Item::Command(cmd)) = kind.as_mut() {
        // find a sibling `code block` that we can use as the 'command'
        let node = node.children().enumerate().find_map(|(index, node)| {
            let d = node.data.borrow();
            match d.value {
                NodeValue::CodeBlock(NodeCodeBlock { fenced: true, .. }) => Some((index, d)),
                _ => None,
            }
        });

        if let Some((index, node)) = node {
            if let NodeValue::CodeBlock(code_block) = &node.value {
                let content = std::str::from_utf8(&code_block.literal).unwrap().trim();
                cmd.with_content(content);
                if !code_block.info.is_empty() {
                    let info = std::str::from_utf8(&code_block.info).unwrap().trim();
                    cmd.with_cli_params(info);
                }
                let mut next_range = path.to_vec();
                next_range.push(index);
                cmd.ast_range = AstRange::range(&next_range, 1)
            }
        }
    }

    if let Some(Item::Topic(topic)) = kind.as_mut() {
        let mut list: Vec<(&'_ Node<RefCell<Ast>>, Option<&'_ Node<RefCell<Ast>>>)> = vec![];
        node.children().enumerate().for_each(|(_index, node)| {
            let d = node.data.borrow();
            match &d.value {
                NodeValue::Heading(NodeHeading { level: 2, .. }) => {
                    let empty_node: Option<&'_ Node<RefCell<Ast>>> = None;
                    list.push((node, empty_node));
                }
                NodeValue::List(_node_list) => {
                    let last = list.last_mut();
                    if let Some(last) = last {
                        if last.1.is_none() {
                            last.1 = Some(node)
                        }
                    }
                }
                _ => {}
            }
        });
        for (heading, maybe_list) in list {
            if let Some(list) = maybe_list {
                let heading_kind = collect_single_line_text(heading);
                let heading_line_start = heading.data.borrow().start_line;
                let _list_data = list.data.borrow();
                for node in list.children() {
                    let d = node.data.borrow();
                    match &d.value {
                        NodeValue::Item(_list) => {
                            for node in node.children() {
                                let named_ref = collect_single_line_text(node);
                                let item_line_start = node.data.borrow();
                                match heading_kind.as_str() {
                                    "Steps" => topic
                                        .add_step_named_ref(named_ref, item_line_start.start_line),
                                    "Dependencies" => topic
                                        .add_dep_named_ref(named_ref, item_line_start.start_line),
                                    _ => {}
                                }
                            }
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    if let Some(kind) = kind {
        items.push(kind)
    }

    items
}

pub(crate) fn to_items(md: &'_ MdElements<'_>) -> Vec<Item> {
    let mut path = vec![0];
    let items = process_node(&md.root, &mut path);

    // for node in root.children() {
    //     let ast = node.data.borrow();
    //
    //     if let Some(Item::Instruction(Instruction { ast, .. })) = kind.as_mut() {
    //         println!("skipping before it's an instruction");
    //         continue;
    //     }
    //
    //     match &ast.value {
    //         NodeValue::Heading(NodeHeading { level: 1, .. }) => {
    //             let t = collect_single_line_text(node);
    //             let item = parse_inline_kind(&t);
    //             kind = item;
    //             // println!("got header.. = {} ", t);
    //             // println!("got item.. = {:?} ", item);
    //         }
    //         NodeValue::Heading(NodeHeading { level: 2, .. }) => {}
    //         NodeValue::Heading(NodeHeading { level: 3, .. }) => {}
    //         NodeValue::Paragraph => {
    //             // println!("++p");
    //             // let ast = node.data.clone().into_inner();
    //             // println!("--p");
    //         }
    //         NodeValue::HtmlBlock(html_block) => {
    //             let html = std::str::from_utf8(&html_block.literal).unwrap();
    //             // println!("++html",);
    //             // println!("\t\t|{}|", html);
    //             // println!("--html",);
    //         }
    //         NodeValue::CodeBlock(cb @ NodeCodeBlock { fenced: true, .. }) => {
    //             let NodeCodeBlock { literal, info, .. } = cb;
    //
    //             // `trim_end` is here because the comrak parser adds a new-line to the
    //             // end of code blocks, which may be spec-compliant, but it's possibly
    //             // something we'd rather not forget about later
    //             let literal = std::str::from_utf8(&literal).unwrap().trim_end();
    //             let info = std::str::from_utf8(&info).unwrap();
    //
    //             // if info.is_empty() {
    //             //     items.push(Element::code_block_without_params(literal));
    //             // } else {
    //             //     items.push(Element::code_block(literal, Some(info)));
    //             // }
    //         }
    //         _ => {}
    //     }
    // }
    //
    // if let Some(item) = kind {
    //     items.push(item);
    // }

    items
}

pub fn debug_ast(asts: &[Ast]) -> impl Debug {
    let _s = String::new();
    struct AstDebug(Vec<Ast>);
    impl Debug for AstDebug {
        fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
            let items = self
                .0
                .iter()
                .map(|ast| {
                    let name = match ast.value {
                        NodeValue::Document => "Document",
                        NodeValue::FrontMatter(_) => "FrontMatter",
                        NodeValue::BlockQuote => "BlockQuote",
                        NodeValue::List(_) => "List",
                        NodeValue::Item(_) => "Item",
                        NodeValue::DescriptionList => "DescriptionList",
                        NodeValue::DescriptionItem(_) => "DescriptionItem",
                        NodeValue::DescriptionTerm => "DescriptionTerm",
                        NodeValue::DescriptionDetails => "DescriptionDetails",
                        NodeValue::CodeBlock(_) => "CodeBlock",
                        NodeValue::HtmlBlock(_) => "HtmlBlock",
                        NodeValue::Paragraph => "Paragraph",
                        NodeValue::Heading(_) => "Heading",
                        NodeValue::ThematicBreak => "ThematicBreak",
                        NodeValue::FootnoteDefinition(_) => "FootnoteDefinition",
                        NodeValue::Table(_) => "Table",
                        NodeValue::TableRow(_) => "TableRow",
                        NodeValue::TableCell => "TableCell",
                        NodeValue::Text(_) => "Text",
                        NodeValue::TaskItem(_) => "TaskItem",
                        NodeValue::SoftBreak => "SoftBreak",
                        NodeValue::LineBreak => "LineBreak",
                        NodeValue::Code(_) => "Code",
                        NodeValue::HtmlInline(_) => "HtmlInline",
                        NodeValue::Emph => "Emph",
                        NodeValue::Strong => "Strong",
                        NodeValue::Strikethrough => "Strikethrough",
                        NodeValue::Superscript => "Superscript",
                        NodeValue::Link(_) => "Link",
                        NodeValue::Image(_) => "Image",
                        NodeValue::FootnoteReference(_) => "FootnoteReference",
                    };
                    format!("start_line: {}: {}", ast.start_line, name)
                })
                .collect::<Vec<String>>();
            f.debug_list().entries(items).finish()
        }
    }
    AstDebug(asts.to_vec())
}

///
/// Single-line text items are identifiers and follow special rules.
///
pub(crate) fn collect_single_line_text<'a>(node: &'a AstNode<'a>) -> String {
    node.children()
        .filter_map(|n| match &n.data.borrow().value {
            NodeValue::Text(t) => Some(std::str::from_utf8(t).unwrap().to_string()),
            // todo, preserve this information?
            NodeValue::Code(t) => Some(std::str::from_utf8(t).unwrap().to_string()),
            _ => None,
        })
        .collect::<Vec<String>>()
        .join("")
}
