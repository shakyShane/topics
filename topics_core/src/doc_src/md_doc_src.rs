use crate::context::Context;
use crate::doc::DocResult;
use crate::doc_err::DocError;
use crate::doc_src::DocSrcImpl;
use crate::items::Item;

use multi_doc::MultiDoc;

use std::path::PathBuf;
use std::str::FromStr;

#[derive(Debug, Clone, Default, serde::Serialize)]
pub struct MdDocSource {
    pub input_file: Option<PathBuf>,
    pub file_content: String,
    pub doc_src_items: MultiDoc,
}

impl DocSrcImpl for MdDocSource {
    fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        let abs = ctx.join_path(pb);
        let file_str = std::fs::read_to_string(&abs).map_err(|e| DocError::PathRead {
            pb: pb.clone(),
            abs: abs.clone(),
            original: e,
        })?;
        let items = MultiDoc::from_str(&file_str)?;
        let new_self = Self {
            input_file: Some(pb.clone()),
            file_content: file_str,
            doc_src_items: items,
        };
        Ok(new_self)
    }
}

impl FromStr for MdDocSource {
    type Err = DocError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let items = MultiDoc::from_md_str(&s)?;
        Ok(Self {
            input_file: None,
            file_content: s.to_string(),
            doc_src_items: items,
        })
    }
}

impl<'a> From<&'a MdDocSource> for &'a str {
    fn from(mds: &'a MdDocSource) -> Self {
        &mds.file_content
    }
}

// pub fn parse_elements(elements: &[Element]) -> DocResult<Vec<Item>> {
//     let mut items: Vec<Item> = vec![];
//     let mut kind: Option<Item> = None;
//     let mut prev_heading2: Option<Element> = None;
//     for elem in elements {
//         match elem {
//             Element::Heading { level: 1, content } => {
//                 let split = content.splitn(2, ':').collect::<Vec<&str>>();
//                 match (split.get(0), split.get(1)) {
//                     (Some(kind_str), Some(rest)) => {
//                         if let Ok(mut item) = Item::from_str(kind_str) {
//                             item.set_name(rest.trim());
//                             kind = Some(item);
//                         }
//                     }
//                     _ => todo!("invalid title {:?}", split),
//                 }
//             }
//             h @ Element::Heading { .. } => {
//                 prev_heading2 = Some(h.clone());
//             }
//             Element::CodeBlock {
//                 params: Some(p),
//                 content,
//             } => {
//                 if let Some(Item::Command(cmd)) = kind.as_mut() {
//                     cmd.with_content(content);
//                     cmd.with_cli_params(&p);
//                 }
//                 if let Some(Item::DependencyCheck(dep_check)) = kind.as_mut() {
//                     use code_fence::Cmd::*;
//                     match code_fence::parse_code_fence_args(&p) {
//                         Ok(Some(Verify(_))) => {
//                             // we only assign this code block if it has ```shell verify ...
//                             dep_check.verify = content.to_string();
//                         }
//                         Ok(Some(AutoFix(_))) => {
//                             // we only assign this code block if it has ```shell autofix ...
//                             dep_check.autofix = Some(content.to_string());
//                         }
//                         _a => {
//                             todo!("handle parsing code-block inline args")
//                         }
//                     }
//                 }
//             }
//             Element::List { items } => {
//                 if let Some(Element::Heading { content, level: 2 }) = &prev_heading2 {
//                     if let Some(Item::Topic(topic)) = kind.as_mut() {
//                         match content.as_str() {
//                             "Dependencies" | "Dependencies:" => {
//                                 for item in items {
//                                     let parsed = parse_item_from_list_items(&item.0);
//                                     if let Some(item_wrap) = parsed {
//                                         topic.deps.push(item_wrap);
//                                     }
//                                 }
//                             }
//                             "Steps" | "Steps:" => {
//                                 for item in items {
//                                     let parsed = parse_item_from_list_items(&item.0);
//                                     if let Some(item_wrap) = parsed {
//                                         topic.steps.push(item_wrap);
//                                     }
//                                 }
//                             }
//                             _ => {}
//                         }
//                     }
//                 }
//                 // println!("prev heading {:?}", prev_heading2);
//             }
//             _ => {}
//         }
//     }
//     if let Some(item) = kind {
//         items.push(item)
//     }
//     Ok(items)
// }

///
/// There are special semantics for parsing an inline list item
///
/// Things we care about
///
// fn parse_item_from_list_items(items: &[Element]) -> Option<ItemWrap> {
//     // 1. if there's a single text node, this is a named reference
//     if let (1, Some(Element::Text { content })) = (items.len(), items.get(0)) {
//         return parse_inline_kind(content)
//             .map(ItemWrap::Item)
//             .or_else(|| Some(ItemWrap::Named(content.to_string())));
//     }
// if let (1, Some(Element::Paragraph { content })) = (items.len(), items.get(0)) {
//     return parse_inline_kind(content)
//         .map(ItemWrap::Item)
//         .or_else(|| Some(ItemWrap::Named(content.to_string())));
// }

// 2. <kind>: <name> pattern. This should be 1 line of kind + name and then `n` trailing lines.
// if let (_len, Some(Element::Paragraph { content })) = (items.len(), items.get(0)) {
//     match parse_inline_kind(content) {
//         Some(Item::Instruction(mut ins)) => {
//             let mut instruction_lines = vec![];
//             items.iter().skip(1).for_each(|rem| match rem {
//                 Element::Paragraph { content } => instruction_lines.push(content.to_string()),
//                 Element::Text { content } => instruction_lines.push(content.to_string()),
//                 _ => {}
//             });
//             ins.instruction = instruction_lines.join("\n");
//             return Some(ItemWrap::Item(Item::Instruction(ins)));
//         }
//         Some(Item::Command(mut cmd)) => {
//             items.iter().skip(1).for_each(|rem| match rem {
//                 Element::CodeBlock {
//                     content,
//                     params: Some(p),
//                 } => {
//                     cmd.with_content(content);
//                     cmd.with_cli_params(p);
//                 }
//                 Element::CodeBlock { content: _, params } => {
//                     if params.is_none() {
//                         eprintln!(
//                             "TODO - missing inline params on codefence from inline COMMAND item. "
//                         )
//                     }
//                 }
//                 _ => {}
//             });
//             return Some(ItemWrap::Item(Item::Command(cmd)));
//         }
//         Some(Item::DependencyCheck(mut dep_check)) => {
//             items.iter().skip(1).for_each(|rem| match rem {
//                 Element::CodeBlock {
//                     content,
//                     params: Some(p),
//                 } => {
//                     dep_check.with_content(content, p);
//                 }
//                 Element::CodeBlock { content: _, params } => {
//                     if params.is_none() {
//                         eprintln!(
//                             "TODO - missing inline params on codefence from inline DependencyCheck item. "
//                         )
//                     }
//                 }
//                 _ => {}
//             });
//             return Some(ItemWrap::Item(Item::DependencyCheck(dep_check)));
//         }
//         _ => { /* noop */ }
//     }
// }

// None
// }

pub fn parse_inline_kind(input: &str) -> Option<Item> {
    // split by lines first - inline kind names must be a single line
    let lines = input.splitn(2, '\n').collect::<Vec<&str>>();

    // kind + name + other
    if let (Some(first), _maybe_rest) = (lines.get(0), lines.get(1)) {
        match split_first_line(first).as_mut() {
            Some(Item::Instruction(inst)) => {
                return Some(Item::Instruction(inst.clone()));
            }
            Some(Item::Command(cmd)) => {
                return Some(Item::Command(cmd.clone()));
            }
            Some(Item::DependencyCheck(dep_check)) => {
                return Some(Item::DependencyCheck(dep_check.clone()));
            }
            Some(Item::Topic(topic)) => {
                return Some(Item::Topic(topic.clone()));
            }
            Some(_v) => {
                todo!("todo inline list item");
            }
            None => {}
        }
    }

    // kind + name only
    if let (Some(_first), None) = (lines.get(0), lines.get(1)) {
        // todo!("kind + name only in a list, needs to be solved")
    }

    None
}

fn split_first_line(first_line_input: &str) -> Option<Item> {
    let split = first_line_input.splitn(2, ':').collect::<Vec<&str>>();
    match (split.get(0), split.get(1)) {
        (Some(kind_str), Some(rest)) => {
            if let Ok(mut item) = Item::from_str(kind_str) {
                item.set_name(rest.trim());
                Some(item)
            } else {
                None
            }
        }
        _ => None,
    }
}

pub mod code_fence {
    use crate::items::{AutoFixInlineArgs, CommandInlineArgs, VerifyInlineArgs};
    use structopt::StructOpt;

    #[derive(Debug, structopt::StructOpt)]
    pub enum Cmd {
        Command(CommandInlineArgs),
        Verify(VerifyInlineArgs),
        #[structopt(alias = "autofix")]
        AutoFix(AutoFixInlineArgs),
    }
    #[derive(Debug, structopt::StructOpt)]
    pub struct CodeBlock {
        #[structopt(subcommand)]
        pub cmd: Cmd,
    }

    pub(crate) fn parse_code_fence_args(input: &str) -> anyhow::Result<Option<Cmd>> {
        if let Ok(words) = split_args(&input) {
            if words.len() > 1 {
                if let Ok(cb) = Cmd::from_iter_safe(&words) {
                    return Ok(Some(cb));
                }
            }
        }
        Ok(None)
    }

    pub(crate) fn split_args(input: &str) -> anyhow::Result<Vec<String>> {
        shellwords::split(input).map_err(|e| e.into())
    }
}
