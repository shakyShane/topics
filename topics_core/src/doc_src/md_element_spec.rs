use crate::doc_src::{collect_single_line_text, to_items, MdDocSource, MdElements, MdSrc};
use crate::items::{Command, Instruction, Item};
use comrak::nodes::{Ast, NodeValue};
use comrak::{format_html, Arena, ComrakOptions};
use std::convert::TryInto;
use std::path::PathBuf;
use std::str::FromStr;

#[test]
fn test_command() -> anyhow::Result<()> {
    let input = r#"# Command: Run unit tests

A block of text following the title!

```shell command --cwd="./"
echo hello world!
```
        "#;
    let mut md_src = MdSrc::new(MdDocSource::from_str(input)?);
    let elems = md_src.parse().as_ref().expect("parse md");
    let items = elems.as_items()?;
    let first = items.get(0).expect("at least 1 item");
    if let Item::Command(Command {
        name,
        command,
        ast_range,
        ..
    }) = first
    {
        assert_eq!(name, &String::from("Run unit tests"));
        assert_eq!(command, &String::from("echo hello world!"));
        let html = elems.as_html(ast_range);
        assert_eq!(
            r#"<pre><code class="language-shell">echo hello world!
</code></pre>
"#,
            html
        )
    } else {
        unreachable!();
    }
    Ok(())
}

#[test]
fn test_instruction() -> anyhow::Result<()> {
    let input = r#"
# Instruction: Call IT help desk

Lorem ipsum dolor sit *amet*, consectetur adipisicing elit. Accusamus assumenda molestiae natus quaerat rerum sed suscipit tenetur? Aperiam minima, quos. Commodi corporis cupiditate facilis in minus, quae quis quos similique!

More stuff here

<h1>Oops!</h1>

```shell
oh feck
```
        "#;
    let arena = Arena::new();
    let md = MdElements::new(input, &arena);
    let items: Vec<Item> = md.as_items()?;
    let first = items.get(0);
    if let Some(Item::Instruction(inst)) = first {
        assert_eq!(inst.name, String::from("Call IT help desk"));
    } else {
        unreachable!();
    }
    Ok(())
}

#[test]
fn test_steps() -> anyhow::Result<()> {
    let command_input = r#"# Command: Run unit tests

A block of text following the title!

```shell command --cwd="./"
echo hello world!
```
        "#;
    let mut md_src = MdSrc::from_str(command_input)?;
    md_src.doc_src.input_file = Some(PathBuf::from("./command.md"));

    let topic_input = r#"# Topic: Run all tests

## Dependencies

- Access to Azure

## Steps

- Run unit tests
- So something else
        "#;
    let mut md_src_topic = MdSrc::from_str(topic_input)?;
    let elems = md_src_topic.parse().as_ref().expect("parse md");
    let items = elems.as_items()?;
    Ok(())
}

#[test]
fn test_dep_check() -> anyhow::Result<()> {
    let input = r#"
# Dependency Check: Node JS installed globally

Node JS is required and should be on version 12.0

```shell verify
node -v
```
"#;
    let mut md_src = MdSrc::from_str(input)?;
    let elems = md_src.parse().as_ref().expect("parse md");
    let items = elems.as_items()?;
    dbg!(items);
    Ok(())
}
