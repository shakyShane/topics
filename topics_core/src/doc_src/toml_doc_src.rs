use crate::context::Context;
use crate::doc::{Doc, DocResult};
use crate::doc_err::{DocError, Location, LocationError};
use crate::doc_src::DocSrcImpl;
use std::path::PathBuf;
use std::str::FromStr;

#[derive(Debug, Clone, Default)]
pub struct TomlDocSource {
    pub input_file: Option<PathBuf>,
    pub file_content: String,
}

impl DocSrcImpl for TomlDocSource {
    fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        let abs = ctx.join_path(pb);
        let file_str = std::fs::read_to_string(&abs).map_err(|e| DocError::PathRead {
            pb: pb.clone(),
            abs: abs.clone(),
            original: e,
        })?;
        let new_self = Self {
            input_file: Some(pb.clone()),
            file_content: file_str,
        };
        Ok(new_self)
    }
}

impl FromStr for TomlDocSource {
    type Err = DocError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self {
            input_file: None,
            file_content: s.to_string(),
        })
    }
}

pub struct TomlError<'a> {
    pub doc: &'a Doc,
    pub toml_err: toml::de::Error,
}

impl<'a> From<TomlError<'a>> for DocError {
    fn from(TomlError { toml_err, doc }: TomlError<'a>) -> Self {
        let mut err = LocationError {
            input_file_src: doc.source.content().to_string(),
            location: None,
            input_file: doc.source.file(),
            description: toml_err.to_string(),
        };
        if let Some((line, col)) = toml_err.line_col() {
            err.location = Some(Location::LineAndCol { line, column: col });
            err.description = err.to_string()
        }
        DocError::SerdeLocationErr(err)
    }
}
