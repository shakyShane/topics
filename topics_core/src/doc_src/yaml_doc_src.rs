use crate::context::Context;
use crate::doc::{Doc, DocResult};
use crate::doc_err::{DocError, Location, LocationError};
use crate::doc_src::DocSrcImpl;
use multi_doc::{MultiDoc, SingleDoc};
use std::path::PathBuf;
use std::str::FromStr;

lazy_static::lazy_static! {
    static ref RE: regex::Regex = regex::Regex::new("at line (\\d+)").unwrap();
}

#[derive(Debug, Clone, Default)]
pub struct YamlDocSource {
    pub input_file: Option<PathBuf>,
    pub file_content: String,
    pub doc_src_items: MultiDoc,
}

impl DocSrcImpl for YamlDocSource {
    fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        let abs = ctx.join_path(pb);
        let file_str = std::fs::read_to_string(&abs).map_err(|e| DocError::PathRead {
            pb: pb.clone(),
            abs: abs.clone(),
            original: e,
        })?;
        let items = MultiDoc::from_yaml_str(&file_str)?;
        let new_self = Self {
            input_file: Some(pb.clone()),
            file_content: file_str,
            doc_src_items: items,
        };
        Ok(new_self)
    }
}

impl FromStr for YamlDocSource {
    type Err = DocError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let items = MultiDoc::from_yaml_str(&s)?;
        Ok(Self {
            input_file: None,
            file_content: s.to_string(),
            doc_src_items: items,
        })
    }
}

pub fn from_serde_yaml_error(
    doc: &Doc,
    single_doc: &SingleDoc,
    serde_error: &serde_yaml::Error,
) -> DocError {
    let mut err = LocationError {
        input_file_src: doc.source.content().to_string(),
        location: Some(Location::Region {
            line_start: single_doc.line_start + 1,
            line_end: single_doc.line_end,
        }),
        input_file: doc.source.file(),
        description: serde_error.to_string(),
    };
    if let Some(location) = serde_error.location() {
        let real_line = location.line() + single_doc.line_start;
        err.location = Some(Location::LineAndColRegion {
            line: real_line,
            column: location.column(),
            line_start: single_doc.line_start + 1,
            line_end: single_doc.line_end,
        });
        err.description = RE
            .replace_all(
                err.description.as_str(),
                format!("at line {}", real_line).as_str(),
            )
            .to_string()
    }
    DocError::SerdeLocationErr(err)
}

#[cfg(test)]
mod test {

    use crate::context::Context;
    use crate::doc::Doc;
    use crate::doc_src::{DocSource, DocSrcImpl, YamlDocSource};
    use std::path::PathBuf;
    use std::str::FromStr;

    #[test]
    fn test_fixture_file() -> anyhow::Result<()> {
        let ctx = Context::from_vec(&[]);
        let pb = PathBuf::from("../fixtures2/topics.yaml");
        let d = YamlDocSource::from_path_buf(&pb, &ctx)?;
        insta::assert_debug_snapshot!(d);
        Ok(())
    }

    #[test]
    fn test_errors_single() -> anyhow::Result<()> {
        let pb = PathBuf::from("/input-yaml.yml");
        let input = r#"
kind: Topic
name: Run screen shot tests
deps
"#;
        let srcs = YamlDocSource::from_str(input)?;
        let doc = Doc::from_doc_src(&pb, DocSource::Yaml(srcs), &Default::default());
        insta::assert_debug_snapshot!(doc);
        Ok(())
    }

    #[test]
    fn test_errors_multi() -> anyhow::Result<()> {
        let pb = PathBuf::from("/input-yaml.yml");
        let input = r#"---

kind: DependencyCheck
name: global-node
verify: node -v
url: https://www.nodejs.org

---

kind: DependencyCheck
name: global-yarn
verify: yarn -v
url: https://yarn.sh/legacy

---

kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps
"#;
        let srcs = YamlDocSource::from_str(input)?;
        let doc = Doc::from_doc_src(&pb, DocSource::Yaml(srcs), &Default::default());
        insta::assert_debug_snapshot!(doc?.errors);
        Ok(())
    }
}
