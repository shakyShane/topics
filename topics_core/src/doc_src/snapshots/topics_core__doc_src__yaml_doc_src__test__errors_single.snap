---
source: topics_core/src/doc_src/yaml_doc_src.rs
expression: doc

---
Ok(
    Doc {
        source: Yaml(
            YamlDocSource {
                input_file: None,
                file_content: "\nkind: Topic\nname: Run screen shot tests\ndeps\n",
                doc_src_items: MultiDoc {
                    items: [
                        SingleDoc {
                            line_start: 0,
                            line_end: 4,
                            content: "\nkind: Topic\nname: Run screen shot tests\ndeps",
                        },
                    ],
                },
            },
        ),
        items: [],
        errors: [
            SerdeLocationErr(
                LocationError {
                    location: Some(
                        LineAndColRegion {
                            line_start: 1,
                            line_end: 4,
                            line: 4,
                            column: 1,
                        },
                    ),
                    input_file: None,
                    input_file_src: "\nkind: Topic\nname: Run screen shot tests\ndeps\n",
                    description: "while parsing a block mapping, did not find expected key at line 4 column 1",
                },
            ),
        ],
    },
)
